src/Sarif.Multitool/Rules/SARIF1017.InvalidIndex.cs

```diff
@@ -0,0 +1,209 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Sarif.Multitool.Rules
+{
+    public class InvalidIndex : SarifValidationSkimmerBase
+    {
+        private readonly MultiformatMessageString _fullDescription = new MultiformatMessageString
+        {
+            Text = RuleResources.SARIF1017_InvalidIndex
+        };
+
+        public override MultiformatMessageString FullDescription => _fullDescription;
+
+        public override FailureLevel DefaultLevel => FailureLevel.Error;
+
+        /// <summary>
+        /// SARIF1017
+        /// </summary>
+        public override string Id => RuleId.InvalidIndex;
+
+        protected override IEnumerable<string> MessageResourceNames => new string[]
+        {
+            nameof(RuleResources.SARIF1017_Default)
+        };
+
+        protected override void Analyze(Address address, string addressPointer)
+        {
+            ValidateArrayIndex(
+                address.Index,
+                Context.CurrentRun.Addresses,
+                addressPointer,
+                "address",
+                "index",
+                $"runs[{Context.CurrentRunIndex}].addresses");
+
+            ValidateArrayIndex(
+                address.ParentIndex,
+                Context.CurrentRun.Addresses,
+                addressPointer,
+                "address",
+                "parentIndex",
+                $"runs[{Context.CurrentRunIndex}].addresses");
+        }
+
+        protected override void Analyze(Artifact artifact, string artifactPointer)
+        {
+            ValidateArrayIndex(
+                artifact.ParentIndex,
+                Context.CurrentRun.Artifacts,
+                artifactPointer,
+                "artifact",
+                "parentIndex",
+                $"runs[{Context.CurrentRunIndex}].artifacts");
+        }
+
+        protected override void Analyze(ArtifactLocation artifactLocation, string artifactLocationPointer)
+        {
+            ValidateArrayIndex(
+                artifactLocation.Index,
+                Context.CurrentRun.Artifacts,
+                artifactLocationPointer,
+                "artifactLocation",
+                "index",
+                $"runs[{Context.CurrentRunIndex}].artifacts");
+        }
+
+        protected override void Analyze(GraphTraversal graphTraversal, string graphTraversalPointer)
+        {
+            ValidateArrayIndex(
```

Comment:
>ValidateArrayIndex [](start = 12, length = 18)

please double check you've covered all 15 identified cases. Am I correct that I find only 13 here? logicalLocations.parentIndex certainly seems missing, can't find the other, assuming there is another.

Reply:
Absolutely. I have a list of cases to cover. Here's the current status (which I think includes work done after you did your review):
```
[ ] One rule for invalid index
    [x] logicalLocation.parentIndex
    [x] artifactLocation.index
    [x] artifact.parentIndex
    [x] result.ruleIndex
    [x] address.index
    [x] address.parentIndex
    [x] threadFlowLocation.index
    [x] graphTraversal.resultGraphIndex
    [x] graphTraversal.runGraphIndex
    [x] webRequest.index
        [x] result.webRequest
        [x] threadFlowLocation.webRequest
    [x] webResponse.index
        [x] result.webResponse
        [x] threadFlowLocation.webResponse
    [x] resultProvenance.invocationIndex
    [ ] reportingDescriptorReference.index
        [x] result.rule
          [ ] Cover the case where there is a toolComponent reference that affects the array this points into.
        [x] configurationOverride.descriptor
            [x] invocation.ruleConfigurationOverrides
            [x] invocation.notificationConfigurationOverrides
        [ ] reportingDescriptorRelationship.target
        [ ] notification.descriptor
        [ ] notification.associatedRule
        [ ] Use in Taxa
        [ ] Use in ReportingDescriptorRelationship
    [ ] toolComponentReference.index
        [ ] toolComponent.associatedComponent
        [ ] reportingDescriptorReference.toolComponent
```

---
In reply to: [305429284](https://github.com/microsoft/sarif-sdk/pull/1570#discussion_r305429284) [](ancestors = 305429284)

Reply:
Great. Resetting my review to 'reviewing' pending your completion of the work. Carry on! Good stuff.

---
In reply to: [305525771](https://github.com/microsoft/sarif-sdk/pull/1570#discussion_r305525771) [](ancestors = 305525771,305429284)